packages/flutter/lib/src/services/text_formatter.dart

```diff
@@ -246,33 +327,80 @@ class FilteringTextInputFormatter extends TextInputFormatter {
   /// string) because both of the "o"s would be matched simultaneously
   /// by the pattern.
   ///
-  /// Additionally, each segment of the string before, during, and
-  /// after the current selection in the [TextEditingValue] is handled
-  /// separately. This means that, in the case of the "Into the Woods"
-  /// example above, if the selection ended between the two "o"s in
-  /// "Woods", even if the pattern was `RegExp('o+')`, the result
-  /// would be "Int* the W**ds", since the two "o"s would be handled
-  /// in separate passes.
-  ///
-  /// See also [String.splitMapJoin], which is used to implement this
-  /// behavior in both cases.
+  /// The filter may adjust the selection and the composing region of the text
+  /// after applying the text replacement, such that they still cover the same
+  /// text. For instance, if the pattern was `o+` and the last character "s" was
+  /// selected: "Into The Wood|s|", then the result will be "Into The W*d|s|",
+  /// with the selection still around the same character "s" despite that it is
+  /// now the 12th character.
+  ///
+  /// In the case where one end point of the selection (or the composing region)
+  /// is strictly inside the banned pattern (for example, "Into The |Wo|ods"),
+  /// that endpoint will be moved to the end of the replacement string (it will
+  /// become "Into The |W*|ds" if the pattern was `o+` and the original text and
+  /// selection were "Into The |Wo|ods").
   final String replacementString;
 
   @override
   TextEditingValue formatEditUpdate(
     TextEditingValue oldValue, // unused.
     TextEditingValue newValue,
   ) {
-    return _selectionAwareTextManipulation(
-      newValue,
-      (String substring) {
-        return substring.splitMapJoin(
-          filterPattern,
-          onMatch: !allow ? (Match match) => replacementString : null,
-          onNonMatch: allow ? (String nonMatch) => nonMatch.isNotEmpty ? replacementString : '' : null,
-        );
-      },
-    );
+    final _TextEditingValueAccumulator formatState = _TextEditingValueAccumulator(newValue);
+    assert(!formatState.debugFinalized);
+
+    final Iterable<Match> matches = filterPattern.allMatches(newValue.text);
+    Match? previousMatch;
+    for (final Match match in matches) {
+      assert(match.end >= match.start);
```

Comment:
In what case can `match.start == match.end`?  I guess that came up and caused this to throw an error.

Reply:
```dart
final RegExp filter = RegExp('[A-Za-z0-9.@-]*');
print(filter.allMatches('ab&&ca@bcabc').map((m)=> '${m.start} - ${m.end}'));
```
prints `(0 - 2, 2 - 2, 3 - 3, 4 - 12, 12 - 12)` so mismatched characters are represented as an empty match it seems

Reply:
Ah got it, thanks.